mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterActor.java

```diff
@@ -0,0 +1,676 @@
+        setupTaskExecutorStateIfNecessary(registration.getTaskExecutorID());
+        log.info("Request for registering {} with the resource cluster {}", registration.getTaskExecutorID(), this);
+        try {
+            final TaskExecutorID taskExecutorID = registration.getTaskExecutorID();
+            final TaskExecutorState state = taskExecutorStateMap.get(taskExecutorID);
+            boolean stateChange = state.onRegistration(registration);
+            mantisJobStore.storeNewTaskExecutor(registration);
+            if (stateChange) {
+                if (state.isAvailable()) {
+                    taskExecutorsReadyToPerformWork.add(taskExecutorID);
+                }
+                updateHeartbeatTimeout(registration.getTaskExecutorID());
+            }
+            log.info("Successfully registered {} with the resource cluster {}", registration.getTaskExecutorID(), this);
+            sender().tell(Ack.getInstance(), self());
+        } catch (Exception e) {
+            sender().tell(new Status.Failure(e), self());
+        }
+    }
+
+    private void onHeartbeat(TaskExecutorHeartbeat heartbeat) {
+        setupTaskExecutorStateIfNecessary(heartbeat.getTaskExecutorID());
+        try {
+            final TaskExecutorID taskExecutorID = heartbeat.getTaskExecutorID();
+            final TaskExecutorState state = taskExecutorStateMap.get(taskExecutorID);
+            boolean stateChange = state.onHeartbeat(heartbeat);
+            if (stateChange) {
+                if (state.isAvailable()) {
+                    taskExecutorsReadyToPerformWork.add(taskExecutorID);
+                }
+            }
+
+            updateHeartbeatTimeout(heartbeat.getTaskExecutorID());
+            sender().tell(Ack.getInstance(), self());
+        } catch (IllegalStateException e) {
+            sender().tell(new Status.Failure(e), self());
+        }
+    }
+
+    private void onTaskExecutorStatusChange(TaskExecutorStatusChange statusChange) {
+        setupTaskExecutorStateIfNecessary(statusChange.getTaskExecutorID());
+        try {
+            final TaskExecutorID taskExecutorID = statusChange.getTaskExecutorID();
+            final TaskExecutorState state = taskExecutorStateMap.get(taskExecutorID);
+            boolean stateChange = state.onTaskExecutorStatusChange(statusChange);
+            if (stateChange) {
+                if (state.isAvailable()) {
+                    taskExecutorsReadyToPerformWork.add(taskExecutorID);
+                } else {
+                    taskExecutorsReadyToPerformWork.remove(taskExecutorID);
+                }
+            }
+
+            updateHeartbeatTimeout(statusChange.getTaskExecutorID());
+            sender().tell(Ack.getInstance(), self());
+        } catch (IllegalStateException e) {
+            sender().tell(new Status.Failure(e), self());
+        }
+    }
+
+    private void onTaskExecutorAssignmentRequest(TaskExecutorAssignmentRequest request) {
```

Comment:
what happens when the task executor doesn't go from Available to Running after it's been assigned for some time?